src/ray/util/url.cc

```diff
@@ -16,64 +16,103 @@
+#include "ray/util/util.h"
 
-boost::asio::ip::tcp::endpoint parse_ip_tcp_endpoint(const std::string &endpoint,
-                                                     int default_port) {
-  const std::string scheme_sep = "://";
-  size_t scheme_begin = 0, scheme_end = endpoint.find(scheme_sep, scheme_begin);
-  size_t host_begin;
-  if (scheme_end < endpoint.size()) {
-    host_begin = scheme_end + scheme_sep.size();
-  } else {
-    scheme_end = scheme_begin;
-    host_begin = scheme_end;
-  }
-  std::string scheme = endpoint.substr(scheme_begin, scheme_end - scheme_begin);
-  RAY_CHECK(scheme_end == host_begin || scheme == "tcp");
-  size_t port_end = endpoint.find('/', host_begin);
-  if (port_end >= endpoint.size()) {
-    port_end = endpoint.size();
+namespace ray {
+
+std::string endpoint_to_url(
+    const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol> &ep,
+    bool include_scheme) {
+  std::string result;
+  switch (ep.protocol().family()) {
+  case AF_INET:
+  case AF_INET6: {
+    boost::asio::ip::tcp protocol = ep.protocol().family() == AF_INET
+                                        ? boost::asio::ip::tcp::v4()
+                                        : boost::asio::ip::tcp::v6();
+    boost::asio::ip::tcp::endpoint e(protocol, 0);
+    RAY_DCHECK(e.size() == ep.size());
+    memcpy(e.data(), ep.data(), ep.size());
+    std::ostringstream ss;
+    ss << e;
+    result = ss.str();
+    if (include_scheme) {
+      result.insert(0, "tcp://");
+    }
+    break;
   }
-  size_t host_end, port_begin;
-  if (endpoint.find('[', host_begin) == host_begin) {
-    // IPv6 with brackets (optional ports)
-    ++host_begin;
-    host_end = endpoint.find(']', host_begin);
-    if (host_end < port_end) {
-      port_begin = endpoint.find(':', host_end + 1);
-      if (port_begin < port_end) {
-        ++port_begin;
-      } else {
-        port_begin = port_end;
-      }
-    } else {
-      host_end = port_end;
-      port_begin = host_end;
+  case AF_UNIX: {
+    size_t path_offset = offsetof(sockaddr_un, sun_path);
+    result.append(reinterpret_cast<const char *>(ep.data()) + path_offset,
+                  ep.size() - path_offset);
+    if (include_scheme) {
+      result.insert(0, "unix://");
     }
-  } else if (std::count(endpoint.begin() + static_cast<ptrdiff_t>(host_begin),
-                        endpoint.begin() + static_cast<ptrdiff_t>(port_end), ':') > 1) {
-    // IPv6 without brackets (no ports)
-    port_begin = port_end;
-    host_end = port_begin;
-  } else {
-    // IPv4
-    host_end = endpoint.find(':', host_begin);
-    if (host_end < port_end) {
-      port_begin = host_end + 1;
+    break;
+  }
+  default:
+    RAY_LOG(FATAL) << "unsupported protocol family: " << ep.protocol().family();
+    break;
+  }
+  return result;
+}
+
+boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>
+parse_url_endpoint(const std::string &endpoint, int default_port) {
```

Comment:
This method should be simplified a bunch (a little pseudocode-y):
```
std::string schema;
std::string url;
if (endpoint.startswith("unix://")) {
  schema = "unix";
  url = endpoint.lstrip("unix://")
} else if (endpoint.startswith("/")) {
  // I would only support absolute unix paths.
  schema = "unix";
  url = endpoint;
} else if (endpoint.startswith("tcp://")) {
  schema = "tcp";
  url = endpoint.lstrip("tcp://");
} else {
  schema = "tcp";
  url = endpoint;
}

if (scheme == "unix") {
  // I don't think we need the windows ifdef here? This code path should not be called on windows, right?
  return boost::asio::local::stream_protocol::endpoint(url);
}
if (scheme == "tcp") {
  // The ScanToken stuff, but let's make ScanToken a private method and not encourage use outside of this file.
  // Also let's use std::stoi for the port parsing.
}
```

Reply:
Done, but note that this won't support `scheme:path` anymore, only `scheme://path`. [It seems both are valid](https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes?), so I tried to handle both properly—that's part of the reason why it was a bit more complicated than just stripping a prefix.